src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationOptions.cs

```diff
@@ -27,41 +31,77 @@ public LinkedInAuthenticationOptions()
+            ClaimActions.MapCustomJson(ClaimTypes.Surname, user => GetMultiLocaleString(user, "lastName"));
+            ClaimActions.MapCustomJson(Claims.PictureUrl, user
+                => user.SelectTokens("$.profilePicture..elements[*].identifiers[0].identifier").Select(u => u.Value<string>()).LastOrDefault());
+            ClaimActions.MapCustomJson(Claims.PictureUrls, user
+                => string.Join(",", user.SelectTokens("$.profilePicture..elements[*].identifiers[0].identifier").Select(u => u.Value<string>())));
         }
 
+        /// <summary>
+        /// Gets or sets the email address endpoint.
+        /// </summary>
+        public string EmailAddressEndpoint { get; set; }
+
         /// <summary>
         /// Gets the list of fields to retrieve from the user information endpoint.
-        /// See https://developer.linkedin.com/docs/fields/basic-profile for more information.
+        /// See https://docs.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin for more information.
         /// </summary>
         public ISet<string> Fields { get; } = new HashSet<string>
         {
             ProfileFields.Id,
             ProfileFields.FirstName,
             ProfileFields.LastName,
-            ProfileFields.FormattedName,
             LinkedInAuthenticationConstants.EmailAddressField
         };
+
+        /// <summary>
+        /// Gets the MultiLocaleString value.
+        /// First checks if a preferredLocale is returned from the payload, then try to return the value from it.
+        /// Then checks if a value is available for the current UI culture.
+        /// Finally, returns the first localized value.
+        /// See https://docs.microsoft.com/en-us/linkedin/shared/references/v2/object-types#multilocalestring
+        /// </summary>
+        /// <param name="user">The payload returned by the user info endpoint.</param>
+        /// <param name="propertyName">The name of the MultiLocaleString property.</param>
+        /// <returns>The property value.</returns>
+        private string GetMultiLocaleString(JObject user, string propertyName)
+        {
+            if (user[propertyName] == null)
+            {
+                return null;
+            }
+
+            var preferredLocale = user[propertyName]["preferredLocale"];
+            const string localizedKey = "localized";
+
+            if (preferredLocale != null)
+            {
+                var preferredKey = $"{preferredLocale["language"]}_{preferredLocale["country"]}";
+                var preferredLocalizedValue = user[propertyName][localizedKey][preferredKey];
+                if (preferredLocalizedValue != null)
+                {
+                    return preferredLocalizedValue.Value<string>();
+                }
+            }
+
+            var currentUiKey = Thread.CurrentThread.CurrentUICulture.ToString().Replace('-', '_');
```

Comment:
I wonder whether we should put all of them into claims and let the integrator decide, rather than guessing at the behaviour they might want (maybe they don't use the current server thread to make UI decisions)?

Reply:
From my tests, LinkedIn always provides the `preferredLocale` property, but as [it is documented](https://docs.microsoft.com/en-us/linkedin/shared/references/v2/object-types#multilocalestring) as non-required I wanted to implement a fallback, but it will be used as first choice.

The current UI culture seems a pretty good candidate as it is set by AspNetCore localization mechanism to load resource files...
And while I'm thinking about it, I am even wondering if it shouldn't be the first option. The names' localization only makes sense for non-latin characters. Maybe a Chinese guy would have set his preferred locale for his LinkedIn profile in Chinese, but an English ASP.Net application would rather get his name in English?

BTW, I agree that we should put all the names into claims, but I'm not sure how to do it. How put the names and the associated cultures in a claim? And maybe we are lacking another extension method to put several values for a single claim (which I could use for profile pictures, too)?

Reply:
Two immediate thoughts:

  1) Maybe there's a way we could add a `Func<>` property of some kind to the options class that would allow the configuring application to chose the culture and/or the multi-locale string?
  2) I'm not sure about how to add them all. Maybe there's a way the claims actions in the options can be configured to add multiple values? Failing that, they could be enumerated in the handler and added manually for each locale found in the response.

Reply:
I'm working on it, and push it as soon as possible!

Reply:
No rush 😃 - I'll probably not be able to re-review until later this evening now anyway.

Reply:
It's done (with additional tests to cover it) 🙂